File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteScheduleRecord.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.krystianwsul.checkme.firebase.DatabaseWrapper;
import com.krystianwsul.checkme.firebase.json.ScheduleWrapper;

public abstract class RemoteScheduleRecord extends RemoteRecord {
    public static final String SCHEDULES = "schedules";

    @NonNull
    private final String mId;

    @NonNull
    private final RemoteTaskRecord mRemoteTaskRecord;

    @NonNull
    final ScheduleWrapper scheduleWrapper;

    RemoteScheduleRecord(@NonNull String id, @NonNull RemoteTaskRecord remoteTaskRecord, @NonNull ScheduleWrapper scheduleWrapper) {
        super(false);

        mId = id;
        mRemoteTaskRecord = remoteTaskRecord;
        this.scheduleWrapper = scheduleWrapper;
    }

    RemoteScheduleRecord(@NonNull RemoteTaskRecord remoteTaskRecord, @NonNull ScheduleWrapper scheduleWrapper) {
        super(true);

        mId = DatabaseWrapper.INSTANCE.getScheduleRecordId(remoteTaskRecord.getProjectId(), remoteTaskRecord.getId());
        mRemoteTaskRecord = remoteTaskRecord;
        this.scheduleWrapper = scheduleWrapper;
    }

    @NonNull
    @Override
    protected ScheduleWrapper getCreateObject() {
        return scheduleWrapper;
    }

    @NonNull
    public String getId() {
        return mId;
    }

    @NonNull
    @Override
    protected String getKey() {
        return mRemoteTaskRecord.getKey() + "/" + SCHEDULES + "/" + mId;
    }

    public abstract long getStartTime();

    @Nullable
    public abstract Long getEndTime();

    @NonNull
    public String getProjectId() {
        return mRemoteTaskRecord.getProjectId();
    }

    @NonNull
    public String getTaskId() {
        return mRemoteTaskRecord.getId();
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteScheduleRecord.kt
package com.krystianwsul.checkme.firebase.records


import com.krystianwsul.common.firebase.ScheduleWrapper
import com.krystianwsul.common.utils.RemoteCustomTimeId

abstract class RemoteScheduleRecord<T : RemoteCustomTimeId> : RemoteRecord {

    companion object {

        const val SCHEDULES = "schedules"
    }

    val id: String

    protected val remoteTaskRecord: RemoteTaskRecord<T>

    final override val createObject: ScheduleWrapper

    override val key get() = remoteTaskRecord.key + "/" + SCHEDULES + "/" + id

    abstract val startTime: Long

    abstract var endTime: Long?

    val projectId get() = remoteTaskRecord.projectId

    val taskId get() = remoteTaskRecord.id

    abstract val customTimeId: RemoteCustomTimeId?

    constructor(id: String, remoteTaskRecord: RemoteTaskRecord<T>, scheduleWrapper: ScheduleWrapper) : super(false) {
        this.id = id
        this.remoteTaskRecord = remoteTaskRecord
        this.createObject = scheduleWrapper
    }

    constructor(remoteTaskRecord: RemoteTaskRecord<T>, scheduleWrapper: ScheduleWrapper) : super(true) {
        id = remoteTaskRecord.getScheduleRecordId()
        this.remoteTaskRecord = remoteTaskRecord
        this.createObject = scheduleWrapper
    }
}
